### PR TITLE
Custom Error messages 

### PR DESCRIPTION
**Pull request by brindasasi**

### CODE REVIEW

File: api/app/Exceptions/CustomException.php [hunk 1]
```diff
@@ -0,0 +1,59 @@
+<?php
+
+namespace App\Exceptions;
+
+use Exception;
+use Nuwave\Lighthouse\Exceptions\RendersErrorsExtensions;
+
+class CustomException extends Exception implements RendersErrorsExtensions
```
Comment (patcon) [inline api/app/Exceptions/CustomException.php:8]:
```suggestion
// Source: https://lighthouse-php.com/master/digging-deeper/error-handling.html#additional-error-information
class CustomException extends Exception implements RendersErrorsExtensions
```

**Comment by petertgiles** [comment]
Thanks for the demo!  Here are my thoughts for next steps:
1. Issue to set up validation errors
  - Create validation rule with error code
  - Suggest conventions for unique, understandable error codes
  - Create a frontend class to collect localized error messages keyed to error codes
  - Create a helper function to take a error code and turn it into a localized error message for toasting
  - Unit test with English and French validation
 2. Issue to set up custom exception exceptions
 - Create a custom validation rule with error code
 - Add custom exception error code to error code class
 - Update helper function to recognize custom exceptions
 - Unit test with English and French validation
3. Issue to make intl-extract smart enough to find error codes to translation

File: api/graphql/schema.graphql [hunk 2]
```diff
@@ -1164,7 +1170,7 @@ input GenericJobTitleBelongsToMany {
 }
 
 type Mutation {
-    createUser(user: CreateUserInput! @spread): User @create @guard @can(ability: "create")
+    createUser(user: CreateUserInput! @spread): User  @guard @can(ability: "create") @field(resolver: "CreateUser")
```
Comment (patcon) [inline api/graphql/schema.graphql:1173]: Is it still possible to leave `@create` there? Or does it need to be reimplemented in CreateUsers?

File: api/graphql/schema.graphql [hunk 3]
```diff
@@ -721,7 +721,13 @@ input CreateUserInput {
...
     firstName: String! @rename(attribute: "first_name")
     lastName: String! @rename(attribute: "last_name")
-    email: Email @rules(apply: ["unique:users,email"])
+    email: Email @rules(apply: ["unique:users,email"],
+    messages: [
+        {
+            rule: "unique"
+            message: "Email address already exists"
+        }
+    ])
```
Comment (patcon) [inline api/graphql/schema.graphql:730]: Even though it's just a spike, to side-step confusion for anyone else reviewing :)

```suggestion
    email: Email @rules(
        apply: ["unique:users,email"],
        messages: [{
            rule: "unique"
            message: "Email address already exists"
        }]
    )
```

File: api/graphql/schema.graphql (hunk 2 above)
Comment (brindasasi) [reply api/graphql/schema.graphql:1173]: they are mutually exclusive
for validation exception -> @create ( default resolver would throw the exception) 
for custom exception -> @field(resolver: "CreateUser")  ( we use custom resolvers to manually throw the exception )

**Comment by brindasasi** [comment]
> Thanks for the demo! Here are my thoughts for next steps:
> 
> 1. Issue to set up validation errors
> 
> * Create validation rule with error code
> * Suggest conventions for unique, understandable error codes
> * Create a frontend class to collect localized error messages keyed to error codes
> * Create a helper function to take a error code and turn it into a localized error message for toasting
> * Unit test with English and French validation
> 
> 2. Issue to set up custom exception exceptions
> 
> * Create a custom validation rule with error code
> * Add custom exception error code to error code class
> * Update helper function to recognize custom exceptions
> * Unit test with English and French validation
> 
> 3. Issue to make intl-extract smart enough to find error codes to translation

Thank you @petertgiles 
Here are the issues 
#3322 
#3328 
#3330

File: api/graphql/schema.graphql (hunk 2 above)
Comment (patcon) [reply api/graphql/schema.graphql:1173]: Thanks Brinda!

**Pull request closed without merging**